pls/functions.py

```diff
@@ -0,0 +1,515 @@
+get_property(VALUE GLOBAL PROPERTY "HAS_SINGLETON_LIBRARY_{lib_name_uppercase}")
+if(NOT VALUE)
+  set_property(GLOBAL PROPERTY "HAS_SINGLETON_LIBRARY_{lib_name_uppercase}" TRUE)
+  add_subdirectory(impl)
+endif()
+"""
+
+
+def pls_fail(msg):
+    # TODO(dkorolev): Write the failure message to an env-provided file, for "unit"-testing `pls`.
+    print(msg)
+    sys.exit(1)
+
+
+if os.path.isfile("pls.py") or os.path.isfile("pls"):
+    pls_fail(
+        "PLS: You are probably running `pls` from the wrong directory. Navigate to your project directory first."
+    )
+
+modules = {}
+
+
+@dataclass
+class PerDirectoryStatus:
+    deps: set = field(default_factory=lambda: set())
+    project_name: str = "pls_project"
+    executables: dict = field(default_factory=dict)
+    executable_deps: defaultdict = field(default_factory=lambda: defaultdict(set))
+    add_to_gitignore = []
+    need_cmakelists_txt: bool = False
+
+
+per_dir = defaultdict(PerDirectoryStatus)
+
+full_abspath = os.path.abspath(".")
+per_dir[full_abspath].add_to_gitignore.append(flags.dotpls)
+per_dir[full_abspath].add_to_gitignore.append(".debug")
+per_dir[full_abspath].add_to_gitignore.append(".release")
+
+
+def create_symlink_with_cmakelists_txt(dst_dir, lib_name, lib_cloned_dir):
+    # TODO(dkorolev): Creating the symlink should involve the `.gitignore` magic.
+    final_dst_path = os.path.join(dst_dir, lib_name)
+    if not os.path.isdir(final_dst_path):
+        wrapper_top_dir = os.path.join(flags.dotpls, "singleton_deps")
+        # TODO(dkorolev): So, the top-level symlinks created can be relative paths, the rest should be absolute. Fix this.
+        wrapper_dir = os.path.join(os.path.abspath(wrapper_top_dir), lib_name)
+        os.makedirs(wrapper_dir, exist_ok=True)
+        wrapper_dir_impl = os.path.join(wrapper_dir, "impl")
+        if not os.path.isdir(wrapper_dir_impl):
+            os.symlink(
+                os.path.abspath(lib_cloned_dir),
+                wrapper_dir_impl,
+                target_is_directory=True,
+            )
+        os.symlink(wrapper_dir, final_dst_path, target_is_directory=True)
+        cmakelists_path = os.path.join(wrapper_dir, "CMakeLists.txt")
+        if not os.path.isfile(cmakelists_path):
+            with open(cmakelists_path, "w") as file:
+                file.write(singleton_cmakelists_txt_contents(lib_name))
+
+
+already_traversed_src_dirs = set()
+
+
+def traverse_source_tree(src_dir="."):
+    # TODO(dkorolev): Traverse recursively.
+    # TODO(dkorolev): `libraries`? And a command to `run` them, if only with `objdump -s`?
+    queue_list = deque()
+    queue_set = set()
+
+    def add_to_queue(src_dir):
+        abs_src_dir = os.path.abspath(src_dir)
+        queue_list.append(abs_src_dir)
+        queue_set.add(abs_src_dir)
+
+    add_to_queue(src_dir)
+    while queue_list:
+        src_dir = queue_list.popleft()
+        if src_dir not in already_traversed_src_dirs and (
```

Comment:
Why break lines? Isn't there some `cols: 120` option in the tool you use for `fmt`?

Reply:
https://chatgpt.com/share/0c6db2a5-9a0f-4cff-a412-34d0708bc24c =)

Reply:
change black line-length=120